src/events/message/commands/fun/rankup.ts

```diff
@@ -0,0 +1,184 @@
+  "4dan": 718593,
+  "5dan": 1953551,
+  "6dan": 5310860,
+  "7dan": 14437910,
+  "8dan": 39250354,
+};
+
+const RANKPOINTS: { [rank: string]: number } = {
+  "8kyu": 2,
+  "7kyu": 3,
+  "6kyu": 8,
+  "5kyu": 21,
+  "4kyu": 55,
+  "3kyu": 149,
+  "2kyu": 404,
+  "1kyu": 1097,
+};
+
+const capitalise = (s: string): string =>
+  s.length ? s[0].toUpperCase() + s.slice(1).toLowerCase() : "";
+
+const getLang = (lang: string): string => LANGSYNONYMS[lang.toLowerCase()] || lang.toLowerCase();
+
+// Retrieve a users language score, or overall score if language is undefined
+const getUserLangScore = async (user: string, lang?: string): Promise<number | undefined> => {
+  return await fetch("https://www.codewars.com/api/v1/users/" + user)
+    .then((response: any): any => (response.status === 404 ? { success: false } : response.json()))
+    .then((result: any): any => {
+      if (result.success === false) return;
+      return (lang ? result.ranks.languages[getLang(lang)] : result.ranks.overall)?.score ?? 0;
+    });
+};
+
+export default async (message: Message, args: CommandArg[], opts: Record<string, string>) => {
+  // Helper function
+  const send = (msg: string) => {
+    message.channel.send(msg);
+  };
+  let username: string;
+  if (args.length == 0) {
+    if (!message.member?.displayName) return;
+    username = message.member.displayName;
+  } else {
+    // Input validation
+    const result = z.tuple([word()]).safeParse(args);
+    if (!result.success) {
+      message.reply(USAGE);
+      return;
+    }
+    [username] = result.data;
+  }
+  // Get mode
+  const mode: Mode = (<any>Mode)[capitalise(String(opts.mode).toLowerCase())] ?? DEFAULTMODE;
+
+  // Get user data
+  const score: number | undefined = await getUserLangScore(username, opts.language);
+  if (score == undefined)
+    return send(`${username} is doomed to stay at :9kyu: forever. (user not found)`);
+
+  // Get target score and rank/user string
+  let nextRank: string;
+  let remaining: number;
+
+  // If target is a rank
+  if (/[1-8](?:kyu|dan)/.test(opts.target)) {
+    [nextRank, remaining] = ["reach `" + opts.target + "`", RANKTHRESHOLDS[opts.target]];
+    if (remaining <= score) return send("Target has already been reached");
+  }
+
+  // If target is a user
+  else if (opts.target) {
+    const targScore: number | undefined = await getUserLangScore(opts.target, opts.language);
+    if (targScore === undefined) return send("Target user could not be found");
+    if (targScore < score) return send(`${username} has already reached ${opts.target}'s rank'`);
+    [nextRank, remaining] = ["overtake " + opts.target, targScore - score + 1];
+  }
+
+  // Otherwise take next rank above user's current
+  else {
+    [nextRank, remaining] = Object.entries(RANKTHRESHOLDS)
+      .map(([rank, points]): [string, number] => [rank, points - score])
+      .filter(([, v]) => v > 0)
+      .reduce((a, c) => (a[1] < c[1] ? a : c));
+    nextRank = "reach `" + nextRank + "`";
+  }
```

Comment:
I'd extract this to a separate function that returns `[nextRank: string, remaining: number]`. The error cases can be done by throwing an error with message used by `send`, or changing the return type to include error message string (`[nextRank: string, remaining: number] | string`).